### PR TITLE
hlc: further simplify (Timestamp).String()

### PR DESCRIPTION
**Pull request by knz**
In my previous iteration of this code I got rid of a dynamic
allocation by observing that the number of decimal digits was
fixed (9) so we just had to know how many digits to print (base-10
logarithm) to know where to start:

```go
buf = append(buf, '.')
const zeroes = "000000000"
buf = append(buf, zeroes[:9-ilog10.NumUint32DecimalDigits(uint32(ns))]...)
buf = strconv.AppendUint(buf, ns, 10)
```

However I was being too smart for my own good. This is really
"algorithm 101": AppendUint can be rewritten to do the job over 9
digits directly instead of stopping when it encounters the head digit:

```go
buf = append(buf, '.', '0', '0', '0', '0', '0', '0', '0', '0', '0')
for i := 0; i < 9; i++ {
  buf[len(buf)-1-i] = byte('0' + ns%10)
  ns = ns / 10
}
```

This is not just readable; it is actually faster.

```
name                old time/op    new time/op    delta
TimestampString-32     104ns ±10%      84ns ± 9%  -18.60%  (p=0.000 n=10+9)
```

Finally by unrolling the loop, we shave 14% on top of that:
```
name                old time/op    new time/op    delta
TimestampString-32     104ns ±10%      72ns ± 5%  -30.09%  (p=0.000 n=10+8)
```

A minor advantage of this patch is that it drops the external
dependency on `go-ilog10` which we were not using otherwise.

Release note: None

### CODE REVIEW

**Review by nvb — APPROVED** [review]
Good idea!

Here are two small tweaks that I'll leave for your consideration. The eliminates the bounds checks for each of these slice assignments and the second eliminates a call to `append` (which also incurs a bounds check).
```go
diff --git a/pkg/util/hlc/timestamp.go b/pkg/util/hlc/timestamp.go
index a2ea27cb1a..f9c0d0cd51 100644
--- a/pkg/util/hlc/timestamp.go
+++ b/pkg/util/hlc/timestamp.go
@@ -49,7 +49,7 @@ func (t Timestamp) String() string {
 
 	w := t.WallTime
 	if w == 0 {
-		buf = append(buf, '0')
+		buf = append(buf, '0', ',')
 	} else {
 		if w < 0 {
 			w = -w
@@ -59,32 +59,34 @@ func (t Timestamp) String() string {
 		s, ns := uint64(w/1e9), uint64(w%1e9)
 		buf = strconv.AppendUint(buf, s, 10)
 
-		buf = append(buf, '.', '0', '0', '0', '0', '0', '0', '0', '0', '0')
+		prev := len(buf)
+		buf = append(buf, '.', '0', '0', '0', '0', '0', '0', '0', '0', '0', ',')
 		// The following lines expand to:
 		// 		for i := 0; i < 9; i++ {
-		// 			buf[len(buf)-1-i] = byte('0' + ns%10)
+		// 			buf[len(buf)-2-i] = byte('0' + ns%10)
 		// 			ns = ns / 10
 		// 		}
-		d := len(buf) - 1
-		buf[d] = byte('0' + ns%10)
-		d, ns = d-1, ns/10
-		buf[d] = byte('0' + ns%10)
-		d, ns = d-1, ns/10
-		buf[d] = byte('0' + ns%10)
-		d, ns = d-1, ns/10
-		buf[d] = byte('0' + ns%10)
-		d, ns = d-1, ns/10
-		buf[d] = byte('0' + ns%10)
-		d, ns = d-1, ns/10
-		buf[d] = byte('0' + ns%10)
-		d, ns = d-1, ns/10
-		buf[d] = byte('0' + ns%10)
-		d, ns = d-1, ns/10
-		buf[d] = byte('0' + ns%10)
-		d, ns = d-1, ns/10
-		buf[d] = byte('0' + ns%10)
+		//
+		// The loop unrolling also permits bounds check elimination.
+		zeroBuf := buf[prev+1 : len(buf)-1]
+		zeroBuf[8] = byte('0' + ns%10)
+		ns = ns / 10
+		zeroBuf[7] = byte('0' + ns%10)
+		ns = ns / 10
+		zeroBuf[6] = byte('0' + ns%10)
+		ns = ns / 10
+		zeroBuf[5] = byte('0' + ns%10)
+		ns = ns / 10
+		zeroBuf[4] = byte('0' + ns%10)
+		ns = ns / 10
+		zeroBuf[3] = byte('0' + ns%10)
+		ns = ns / 10
+		zeroBuf[2] = byte('0' + ns%10)
+		ns = ns / 10
+		zeroBuf[1] = byte('0' + ns%10)
+		ns = ns / 10
+		zeroBuf[0] = byte('0' + ns%10)
 	}
-	buf = append(buf, ',')
 	buf = strconv.AppendInt(buf, int64(t.Logical), 10)
 
 	return *(*string)(unsafe.Pointer(&buf))
```

```
name                old time/op    new time/op    delta
TimestampString-16    58.5ns ± 6%    54.5ns ± 6%  -6.79%  (p=0.050 n=10+10)

name                old alloc/op   new alloc/op   delta
TimestampString-16     32.0B ± 0%     32.0B ± 0%    ~     (all equal)

name                old allocs/op  new allocs/op  delta
TimestampString-16      1.00 ± 0%      1.00 ± 0%    ~     (all equal)
```

> Reviewed 4 of 4 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44392#-:-LzcLnTJ5OrQI8rEm1t9:b-5ucr3r)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz)

___
*[pkg/util/hlc/timestamp_test.go, line 142 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44392#-LzcH1DuDsQnO6JfPbYa:-LzcH1DuDsQnO6JfPbYb:bqrk1x) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8c8ded51cf6f264c17546de20ac96e3a6e2661c/pkg/util/hlc/timestamp_test.go#L142)):*
> ```Go
> 
> 	for i := 0; i < b.N; i++ {
> 		unusedStr = ts.String()
> ```

This can just be `_ = ts.String()`

I don't know exactly why, but benchmarks don't seem to perform dead-code elimination.

**Review by knz — COMMENTED** [review]
> Here are two small tweaks that I'll leave for your consideration.

I'll take them, but I'll leave it to you to double-check the performance. On my machine, I get a _slowdown_ of about 4% not a speedup. 

bors r=nvanbenschoten

> Reviewed 1 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44392#-:-LzfS7v94IoysrtQwgyz:b-ud2thc)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten)

**Comment by knz** [comment]
unrelated error

bors r=nvanbenschoten

**Comment by knz** [comment]
The gods are unhappy today.

**Comment by knz** [comment]
waiting on #44437 to merge

**Comment by knz** [comment]
bors r=nvanbenschoten

**Pull request merged**
